hearth/lib/hearth/connection_pool.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+module Hearth
+  # @api private
+  class ConnectionPool
+    @pools_mutex = Mutex.new
+    @pools = {}
+
+    class << self
+      # @return [ConnectionPool]
+      def for(config = {})
+        @pools_mutex.synchronize do
+          puts "using connection pool for #{config}\n" if @pools[config]
+          puts "creating connection pool for #{config}\n" unless @pools[config]
+          @pools[config] ||= new
+        end
+      end
+
+      # @return [Array<ConnectionPool>] Returns a list of the
+      #   constructed connection pools.
+      def pools
+        @pools_mutex.synchronize do
+          @pools.values
+        end
+      end
+    end
+
+    # @api private
+    def initialize
+      @pool_mutex = Mutex.new
+      @pool = {}
```

Comment:
A ConnectionPool having a pool is a little weird to me - esp because a pool here is actually a map of endpoint to pool (so its multiple pools.  So a ConenctionPool has pools?)

Reply:
pool is the internal data store of a ConnectionPool instance. The class method works across multiple pools, which makes sense IMO. I think it depends on how you define a pool. To me it's more than the array portion. It's every connection related to the client's configuration.

Reply:
To me its still just weird to have ConnectionPool class have a `pools` array of ConnectionPool.  Then each ConnectionPool has a hash or `pools` where the values are arrays.  `pools` is just being overloaded and its confusing to have `pools` on both the class and instance level to me.

Reply:
From an active model-esque perspective, I think it's fairly idiomatic.. because ConnectionPool class methods interact with all pools, and instances interact with one pool. Each pool has 1 or many endpoints for one specific config. Multiple pools exist across other configs and potentially other http clients. I can see how it differs a bit if pool is just an array of connections.